File: aoc23/day2/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	file, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Println("failed to read file")
	}

	println(solve(string(file)))
}

func solve(data string) int {

	lines := strings.Split(strings.TrimSpace(data), "\n")
	total := 0

	// colorMap := map[string]int{
	// 	"red":   12,
	// 	"green": 13,
	// 	"blue":  14,
	// }

	for _, line := range lines {
		game := strings.Split(line, ":")[1]
		rolls := strings.Split(game, ";")

		// lineValid := true
		red, blue, green := 0, 0, 0

		for _, roll := range rolls {
			colors := strings.Split(roll, ",")

			for _, c := range colors {
				temp := strings.Split(strings.TrimSpace(c), " ")
				numStr, col := temp[0], temp[1]
				num, _ := strconv.Atoi(numStr)

				switch col {
				case "red":
					red = max(red, num)
				case "green":
					green = max(green, num)
				case "blue":
					blue = max(blue, num)
				}
			}
		}

		total += red * green * blue
	}

	return total
}

File: aoc23/day2/src/main.rs
use std::{cmp, fs, io::Read};

fn main() {
    let mut file = fs::File::open("input.txt").unwrap();
    let mut data = String::new();
    let _ = file.read_to_string(&mut data);


    let output = solve(&data);
    println!("{output}");
}

fn solve(data: &str) -> u32 {
    // let mut ref_map = HashMap::new();
    // ref_map.insert("red", 12);
    // ref_map.insert("green", 13);
    // ref_map.insert("blue", 14);

    return data
        .lines()
        .map(|line| {
            // let mut valid = true;
            let mut splits = line.split(":");
            splits.next();
            let games = splits.next().unwrap();
            let (mut red, mut green, mut blue) = (0, 0, 0);

            for game in games.split(";") {
                for set in game.split(",") {
                    println!("{set}");
                    let mut temp = set.trim().split(" ");
                    let number: u32 = temp.next().unwrap().parse().unwrap();
                    let color = temp.next().unwrap();

                    // match ref_map.get(&color).unwrap().cmp(&number) {
                    //     std::cmp::Ordering::Less => {
                    //         valid = false;
                    //         break;
                    //     }
                    //     _ => (),
                    // }
                    match color {
                        "red" => {
                            red = cmp::max(red, number);
                        }
                        "green" => {
                            green = cmp::max(green, number);
                        }
                        "blue" => {
                            blue = cmp::max(blue, number);
                        }
                        _ => (),
                    }
                }
                // if !valid { break }
            }

            // return if valid { idx as u32 + 1 } else { 0 };
            return red * green * blue;
        })
        .sum();
}
